SilKit/source/core/vasio/RingBuffer.hpp

```diff
@@ -0,0 +1,207 @@
+// SPDX-FileCopyrightText: 2023 Vector Informatik GmbH
+//
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <vector>
+#include <iostream>
+#include <algorithm>
+#include <cstring>
+
+#include "silkit/participant/exception.hpp"
+
+namespace SilKit {
+namespace Core {
+
+template <typename T>
+class RingBuffer
+{
+public:
+    // types
+    using BufArray = std::pair<T*, size_t>;
+
+public:
+    // constructors and destructors
+    RingBuffer(size_t capacity);
+
+public:
+    // public member fcns
+    size_t Capacity();
+    size_t Size();
+
+    bool Read(std::vector<T>& elem, bool advanceRPos = true);
+
+    // write perspective (arrays of free memory in ring buffer)
+    auto GetArrayOne() -> BufArray; // first physically contiguous array (starting at index _wPos)
```

Comment:
Could we hide `GetArrayOne` and `GetArrayTwo` behind a `Write(size_t numBytes) -> BufArray` method? it could be called multiple times and return an invalid `{nullptr, 0}` BufArray when buffer space is unavailable.

Reply:
Regarding the ring buffer vs double buffering issue: 
I got inspired by the Boost circular_buffer (https://www.boost.org/doc/libs/1_85_0/doc/html/boost/circular_buffer.html) with the arrayOne/arrayOne notation.
![image](https://github.com/user-attachments/assets/a9b90d8f-039f-4edd-afba-46a01e3958f1)

Reply:
I like the idea of hiding the arrayOne/arrayTwo objects in a 'Write' method. Unfortunately, writing in the VAsioPeer is not performed directly, rather done implicitely by Asio (in the async_read_some method). 

Therefore, I suggest to hide the details in 'ReadSomeAsync' in a method 'SetUpReceivingBuffersForWrite(std::vector<MutableBuffer> _currentReceivingBuffers& buffers)' - or a similar name.

Reply:
I think giving it a simple name like `auto GetWritingBuffers() -> vector<BufArray>` would be enough to make it obvious (untested pseudo code follows):
```c++
void VAsioPeer::ReadSomeAsync()
{
    SILKIT_ASSERT(_msgBuffer.Capacity() - _msgBuffer.Size() > 0);

    _currentReceivingBuffers.clear();

    auto bufferSequence = _msgBuffer.GetWritingBuffers();
    _currentReceivingBuffers.resize(bufferSequence.size());
    // convert it into asio data structs, or MutableBuffers
    std::transform(bufferSequence.begin(), bufferSequence.end(), _currentReceivingBuffers.begin(),
                       [](const RingBuffer::BufArray& buffer) -> MutableBuffer {
            return MutableBuffer{buffer.first, buffer.second};
    });
    _socket->AsyncReadSome(MutableBufferSequence{_currentReceivingBuffers.data(), _currentReceivingBuffers.size()});
}
```
it could return one of `{}` , `{ArrayOne}`, or `{ArrayOne, ArrayTwo}`. Would that work?

Also it might make sense to re-use MutableBuffer for use in the RingBuffer return values for ArrayOne and Two.